ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java

```diff
@@ -706,46 +765,132 @@ private void processNewPut(ScanResults results, IndexValue putValue) {
         handleContainerBucketUpdate(results, putValue, expiresAtMs, SUBTRACT);
       }
     }
-    if (!isExpired(expiresAtMs, results.logSegmentForecastStartTimeMs)) {
+    if (!isExpired(expiresAtMs, results.logSegmentForecastStartTimeMsForExpired)) {
       results.updateLogSegmentBaseBucket(putValue.getOffset().getName(), putValue.getSize());
       if (expiresAtMs != Utils.Infinite_Time) {
-        handleLogSegmentBucketUpdate(results, putValue, expiresAtMs, SUBTRACT);
+        handleLogSegmentExpiredBucketUpdate(results, putValue, expiresAtMs, SUBTRACT);
       }
     }
   }
 
   /**
    * Helper function to process new DELETE entries and make appropriate updates to the given {@link ScanResults}.
    * @param results the {@link ScanResults} to apply the updates to
+   * @param key the {@link StoreKey} of the new DELETE
    * @param deleteValue the {@link IndexValue} of the new DELETE
    * @param originalPutValue the {@link IndexValue} of the original PUT that is getting deleted
+   * @param previousValue the {@link IndexValue} comes prior to the new DELETE
    */
-  private void processNewDelete(ScanResults results, IndexValue deleteValue, IndexValue originalPutValue) {
+  private void processNewDelete(ScanResults results, StoreKey key, IndexValue deleteValue, IndexValue originalPutValue,
+      IndexValue previousValue) {
+    long operationTimeInMs = deleteValue.getOperationTimeInMs();
+    if (operationTimeInMs == Utils.Infinite_Time) {
+      operationTimeInMs =
+          index.getIndexSegments().floorEntry(deleteValue.getOffset()).getValue().getLastModifiedTimeMs();
+    }
+    // First deal with container bucket
+    if (previousValue.isDelete()) {
+      // The previous value is already DELETE, current DELETE would not change container bucket.
+      // Assume the record history looks like this: PUT, DELETE, DELETE.
+      // First DELETE already updated container bucket.
+    } else {
+      // If original PUT expires before this DELETE, then it will not change container bucket.
+      if (!isExpired(originalPutValue.getExpiresAtMs(), operationTimeInMs)) {
+        handleContainerBucketUpdate(results, originalPutValue, operationTimeInMs, SUBTRACT);
+        handleContainerBucketUpdate(results, originalPutValue, originalPutValue.getExpiresAtMs(), ADD);
+      }
+    }
+    // Now deal with log segment bucket
+    // Current DELETE is always valid
+    if (previousValue.isDelete()) {
+      // Previous IndexValue is a DELETE, it should already
+      // 1. Added it's size to the first bucket in the scan result
+      // 2. Subtracted original PUT's size from the previousValue.getOperationTimeInMs() bucket.
+      // We need to recover from these two operations, so
+      // 1. Subtract previous DELETE's size from bucket in the scan result.
+      // 2. Add original PUT's size to the previousValue.getOperationTimeInMs() bucket.
```

Comment:
if we are under this if statement, then we know the current IndexValue is a DELETE and the previous IndexValue is also a DELETE, the IndexValue history can go like this, PUT, DELETE(v0), DELETE(v1).

When we see the DELETE(v0), we will 1) add size of DELETE to base bucket, then 2)subtract size of PUT int the DELETE's operation time bucket. 
When we see the DELETE(v1), we know it will invalidate DELETE(v0). So one way to do this is to revert what DELETE(v0) did before, so we 1)subtract size of DELETE from base bucket, then 3)add size of PUT in the DELETE(v0)'s operation time bucket. After doing this, we can treat DELETE(v0) just like it never happened. Then we move on to deal with DELETE(v1).

Reply:
Correct, this is what I thought about. My suggestion is to make the comments more accurate:
`Added it's size to the first bucket in the scan result` -> `Added it's size to the base bucket`;
`Subtract previous DELETE's size from bucket in the scan result` -> `Subtract previous DELETE's size from base bucket`;
Also add one sentence in the end something like:
```
After recovering changes made by previous DELETE, we update both base bucket and operation time bucket for new DELETE.
```

Reply:
I reword the comments, please take another look, see if it's good for you.